### PR TITLE
Plan file support

### PR DESCRIPTION
**Pull request by chhwang**
* Completely separate model definition from implementation
* Plan file details full implementation
* Models and plans are serialized into JSON
* Deeper integration with MSCCL++
* A few bug fixes

### CODE REVIEW

File: ark/codegen.cpp [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+            auto it = unsynced.begin();
+            while (it != unsynced.end()) {
+                auto &range = *it;
+                auto intersec = proc_range.intersection(range);
+                if (intersec.empty()) {
+                    it++;
+                    continue;
+                }
+                // no need to sync if range is a part of proc_range.
+                if (intersec.size() < range.size()) {
```
Comment (Binyang2014) [inline ark/codegen.cpp:129]: For these code, do you mean we don't need sync if proc_ranges are same for two ops? What if op2 depends on op1 with same proc_range but thread-block with same id deal with different part of data. Is this possible?

File: ark/codegen.cpp [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        //   size_t k = threadIdx.x / warp_size / slot_n_warps;
+        //   size_t task_id_base = task_b + task_s*p*task_gran;
+        //   for (size_t t = k; ; t += n_slots) {
+        //     size_t task_id = task_id_base + task_s*(
+        //       t/task_gran*task_gran*n_procs + t%task_gran
+        //     );
+        //     if (task_id >= task_e) break;
+        //     task_func(_buf, task_id, sram_per_warp);
+        //   }
+        //   __syncthreads();
```
Comment (Binyang2014) [inline ark/codegen.cpp:376]: Will we call `__syncthreads()` after each task? Or we will call `__syncthreads()` when task across resourceGroups

File: ark/model/model_node.cpp [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "model_node.hpp"
+
+namespace ark {}  // namespace ark
```
Comment (Binyang2014) [inline ark/model/model_node.cpp:6]: Why is empty?

**Review by Binyang2014 — COMMENTED** [review]
Overall looks good to me. Just have several questions. Will finish the review soon

File: ark/model/model_buffer.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    j["Rank"] = rank_;
+    Json send_tags = Json::array();
+    Json recv_tags = Json::array();
+    for (const auto &info : send_tags_) {
+        send_tags.push_back({info.first, info.second});
+    }
+    for (const auto &info : recv_tags_) {
+        recv_tags.push_back({info.first, info.second});
+    }
+    j["SendTags"] = send_tags;
```
[the remark this answers is not in the document]
Comment (chhwang) [reply ark/model/model_buffer.cpp:47]: Yes, this is for letting multiple remote ranks access a single local buffer in collective communication.

File: ark/codegen.cpp [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+    size_t num_procs_;
+    size_t num_warps_per_proc_;
+    std::string code_;
+};
+
+CodeGenerator::Impl::Impl(const Json &plan,
+                          const std::map<size_t, size_t> &buffer_id_to_offset,
+                          const std::string &name)
+    : buffer_id_to_offset_(buffer_id_to_offset), name_(name) {
+    rank_ = plan.at("Rank");
```
[the remark this answers is not in the document]
Comment (chhwang) [reply ark/codegen.cpp:82]: Yes, right

File: ark/api/executor.cpp [hunk 6]
```diff
@@ -0,0 +1,799 @@
...
+    void *lss_ptr_addr = get_global_rt("ARK_LOOP_SYNC_STATE");
+    void *buf_ptr_addr = get_global_rt("ARK_BUF");
+    std::array<int, 4> data = {0, 0, 0, 0};
+    GLOG(gpuMemcpyAsync(lss_ptr_addr, data.data(), sizeof(int) * data.size(),
+                        gpuMemcpyHostToDevice, copy_stream_->get()));
+    GLOG(gpuMemcpyAsync(buf_ptr_addr, &buf_ptr_val, sizeof(gpuDeviceptr),
+                        gpuMemcpyHostToDevice, copy_stream_->get()));
+    if (world_size_ > 1) {
+        void *proxy_chan_addr = get_global_rt("ARK_PROXY_CHANS");
+        void *proxy_secondary_chan_addr =
```
[the remark this answers is not in the document]
Comment (chhwang) [reply ark/api/executor.cpp:581]: We want to have two ProxyChannels per rank to leverage backup links.

File: ark/gpu/gpu_kernel_test.cpp [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+const std::string void_kernel = "extern \"C\" __global__ void kernel() {}";
+
+ark::unittest::State test_gpu_kernel() {
+    ark::GpuKernel kernel(0, void_kernel, {1, 1, 1}, {1, 1, 1}, 0, "kernel");
+    kernel.compile();
+    return ark::unittest::SUCCESS;
+}
+
+//
+// const std::string test_kernel_loop_void =
```
[the remark this answers is not in the document]
Comment (chhwang) [reply ark/gpu/gpu_kernel_test.cpp]: Removed comments in [1fbf38b](https://github.com/microsoft/ark/pull/205/commits/1fbf38b4dd839f97199ade30a90959c5145f1a86).

File: ark/codegen.cpp (hunk 1 above)
Comment (chhwang) [reply ark/codegen.cpp:128]: Sorry for confusion, actually we do sync when proc_ranges are the same. I updated comments in [1fbf38b](https://github.com/microsoft/ark/pull/205/commits/1fbf38b4dd839f97199ade30a90959c5145f1a86).

File: ark/codegen.cpp (hunk 2 above)
Comment (chhwang) [reply ark/codegen.cpp]: No, we actually don't call `__syncthreads()` here. I updated comments in [1fbf38b](https://github.com/microsoft/ark/pull/205/commits/1fbf38b4dd839f97199ade30a90959c5145f1a86).

File: ark/model/model_node.cpp (hunk 3 above)
Comment (chhwang) [reply ark/model/model_node.cpp]: Removed this file in [1fbf38b](https://github.com/microsoft/ark/pull/205/commits/1fbf38b4dd839f97199ade30a90959c5145f1a86).

File: ark/api/executor.cpp [hunk 8]
```diff
@@ -0,0 +1,799 @@
...
+        std::vector<size_t> offsets;
+        int len;
+        auto bootstrap = comm_->bootstrap();
+        bootstrap->recv(&len, sizeof(int), remote_rank, 3);
+        tags.resize(len);
+        offsets.resize(len);
+        bootstrap->recv(tags.data(), len * sizeof(int), remote_rank, 4);
+        bootstrap->recv(offsets.data(), len * sizeof(size_t), remote_rank, 5);
+        for (int i = 0; i < len; ++i) {
+            buffer_id_to_offset[recv_tag_to_buffer_id[tags[i]]] = offsets[i];
```
[the remark this answers is not in the document]
Comment (chhwang) [reply ark/api/executor.cpp:461]: If the sending data is on the same buffer, we can use the same tag. I added more explanation in [650f15f](https://github.com/microsoft/ark/pull/205/commits/650f15f54d45b45ae8e1828fdbffa2876d92a8e6).

File: ark/model/model_graph_impl.cpp [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+        }
+        for (auto &consumer : node->consumers) {
+            if (consumer->producers.find(node) == consumer->producers.end()) {
+                LOG(DEBUG, "consumer does not have this node as producer");
+                return false;
+            }
+        }
+    }
+    try {
+        [[maybe_unused]] auto j = Json::parse(this->serialize(-1));
```
Comment (Binyang2014) [inline ark/model/model_graph_impl.cpp:124]: Will the compiler eliminate this sentence if `j` is not used

File: ark/include/kernels/common/unit_op.h [hunk 10]
```diff
@@ -51,14 +51,14 @@ struct UnitOp {
...
 
-    static_assert(_OutDims::N % _UnitOutDims::N == 0,
-                  "Dimension N is not divisible by the unit dimension");
-    static_assert(_OutDims::C % _UnitOutDims::C == 0,
-                  "Dimension C is not divisible by the unit dimension");
-    static_assert(_OutDims::H % _UnitOutDims::H == 0,
-                  "Dimension H is not divisible by the unit dimension");
-    static_assert(_OutDims::W % _UnitOutDims::W == 0,
-                  "Dimension W is not divisible by the unit dimension");
+    // static_assert(_OutDims::N % _UnitOutDims::N == 0,
```
Comment (Binyang2014) [inline ark/include/kernels/common/unit_op.h:54]: No needed?

**Review by Binyang2014 — APPROVED** [review]
LGTM. I don't review all files. Since most of op related code are similar, just reviewed some of them

File: ark/include/kernels/common/unit_op.h (hunk 10 above)
Comment (chhwang) [reply ark/include/kernels/common/unit_op.h]: Updated in [ecd5afd](https://github.com/microsoft/ark/pull/205/commits/ecd5afd22a3626f27ad476e2094d00f294492d89).

File: ark/model/model_graph_impl.cpp (hunk 9 above)
Comment (chhwang) [reply ark/model/model_graph_impl.cpp]: Modified in [ce5d87a](https://github.com/microsoft/ark/pull/205/commits/ce5d87a74087e8dfc1254a4f7c72e6be56881fab) and added a TODO.

File: examples/llama/model_test.py [hunk 11]
```diff
@@ -428,23 +430,43 @@ def test_transformer_block(
...
-            world_size,
-        ],
-        inputs_ark=[feature, 0, freqs_cis_ark, None],
-        module_class_pt=model_pt.TransformerBlock,
-        module_args_pt=[0, args],
-        inputs_pt=[feature.astype(dtype), 0, freqs_cis, None],
-        module_name_prefix="layers.0",
-        rank=rank,
-        world_size=world_size,
+    module = model_ark.Attention(
```
[the remark this answers is not in the document]
Comment (chhwang) [reply examples/llama/model_test.py:433]: Will update this in another PR

**Pull request merged**